src/global/common.rs

```diff
@@ -185,6 +187,138 @@ pub(crate) async fn find_package_records(conda_meta: &Path) -> miette::Result<Ve
+                    "{}Added exposed '{exposed}' for environment '{env}'.",
+                    console::style(console::Emoji("✔ ", "")).green(),
+                )
+            }
+            StateChange::RemovedExposed(exposed, env) => {
+                write!(
+                    f,
+                    "{}Removed exposed '{exposed}' for environment '{env}'.",
+                    console::style(console::Emoji("✔ ", "")).green(),
+                )
+            }
+            StateChange::UpdatedExposed(exposed, env) => {
+                write!(
+                    f,
+                    "{}Updated exposed '{exposed}' for environment '{env}'.",
+                    console::style(console::Emoji("✔ ", "")).green(),
+                )
+            }
+            StateChange::AddedPackage(pkg, env) => write!(
+                f,
+                "{}Added package '{}={}' to environment '{env}'.",
+                console::style(console::Emoji("✔ ", "")).green(),
+                pkg.name.as_normalized(),
+                pkg.version,
+            ),
+            StateChange::AddedEnvironment(env) => write!(
+                f,
+                "{}Added environment '{env}'.",
+                console::style(console::Emoji("✔ ", "")).green()
+            ),
+            StateChange::RemovedEnvironment(env) => write!(
+                f,
+                "{}Removed environment '{env}'.",
+                console::style(console::Emoji("✔ ", "")).green()
+            ),
+        }
+    }
+}
+
+#[must_use]
+#[derive(Debug, Default)]
+pub(crate) struct StateChanges {
+    changes: Vec<StateChange>,
+    has_updated: bool,
+}
+
+impl StateChanges {
+    pub(crate) fn has_changed(&self) -> bool {
+        self.has_updated || !self.changes.is_empty()
+    }
+
+    pub(crate) fn set_has_updated(&mut self, has_updated: bool) {
+        self.has_updated = has_updated;
+    }
+
+    pub(crate) fn push_change(&mut self, change: StateChange) {
+        self.changes.push(change);
+    }
+
+    pub(crate) fn push_changes(&mut self, changes: impl IntoIterator<Item = StateChange>) {
+        self.changes.extend(changes);
+    }
+
+    #[cfg(test)]
+    pub fn changes(self) -> Vec<StateChange> {
+        self.changes
+    }
+
+    /// Remove changes that cancel each other out
+    fn prune(&mut self) {
+        // Remove changes if the environment is removed afterwards
+        let mut pruned_changes: Vec<StateChange> = Vec::new();
+        for change in &self.changes {
+            match change {
+                StateChange::RemovedEnvironment(env) => {
+                    pruned_changes.retain(|change| match change {
+                        StateChange::RemovedEnvironment(_) => true,
+                        c => c.env() != env,
+                    });
+                }
+                _ => {
+                    pruned_changes.push(change.clone());
+                }
+            }
+        }
+        self.changes = pruned_changes;
+    }
```

Comment:
This does little but can be extended.

For example, when implementing `pixi global remove` we could make sure that `RemovedPackage` and `AddedPackage` cancel each other out.